### PR TITLE
Allow injection of functions into composite values, refactor PublicKey based on it

### LINKED ISSUE #2874: Allow declaration of values in specific locations
Needed for EVM work

### PR DESCRIPTION
**Pull request by turbolent**
Closes #2874 

## Description

The host might want composite values of certain types to have native functions. Add a function `SetCompositeValueFunctionsHandler` to the environment which allows that.

Extend the `CompositeValue.GetMember` implementation, from only linking in interpreted functions, to also querying a handler set in the interpreter configuration.

As the built-in `PublicKey` type has native functions (e.g. `verify`) and is already implemented as a `CompositeValue`, refactor its existing implementation from injecting the native functions at construction time, to using the new injection mechanism added in this PR.

This essentially allows making the `PublicKey` type storable (in a follow-up PR), because the native functions get linked in even after a public key value got loaded from storage. `PublicKey` values are not yet storable exactly because of this missing function injection feature.

______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: runtime/interpreter/interpreter.go [hunk 1]
```diff
@@ -4785,6 +4792,77 @@ func (interpreter *Interpreter) GetContractComposite(contractLocation common.Add
...
+
+	var computedFields map[string]ComputedField
+	if v.Location == nil {
+		computedFields = GetNativeCompositeValueComputedFields(v)
+		if computedFields != nil {
+			return computedFields
+		}
+	}
+
+	// TODO: add handler to config
```
Comment (turbolent) [inline runtime/interpreter/interpreter.go:4826]: This is a TODO for a future enhancement, it is not needed atm

File: runtime/interpreter/value.go [hunk 2]
```diff
@@ -16235,8 +16235,8 @@ type CompositeValue struct {
 	Location        common.Location
 	staticType      StaticType
 	Stringer        func(gauge common.MemoryGauge, value *CompositeValue, seenReferences SeenReferences) string
-	InjectedFields  map[string]Value
-	ComputedFields  map[string]ComputedField
+	injectedFields  map[string]Value
+	computedFields  map[string]ComputedField
```
Comment (turbolent) [inline runtime/interpreter/value.go:16239]: Unexport, so they are not used before being loaded

File: runtime/interpreter/value.go [hunk 3]
```diff
@@ -16598,19 +16598,15 @@ func (v *CompositeValue) GetMember(interpreter *Interpreter, locationRange Locat
...
-		if !goerrors.As(err, &keyNotFoundError) {
-			panic(errors.NewExternalError(err))
+	// Give computed fields precedence over stored fields for built-in types
+	if v.Location == nil {
+		if computedField := v.GetComputedField(interpreter, locationRange, name); computedField != nil {
+			return computedField
 		}
 	}
-	if storable != nil {
-		return StoredValue(interpreter, storable, config.Storage)
```
Comment (turbolent) [inline runtime/interpreter/value.go:16613]: Duplicate of `GetField`

File: runtime/interpreter/value_publickey.go [hunk 4]
```diff
@@ -89,32 +71,5 @@ func NewPublicKeyValue(
...
-		}
-
-		return formatComposite(
-			memoryGauge,
-			string(publicKeyValue.TypeID()),
-			stringerFields,
-			seenReferences,
-		)
-	}
-
```
Comment (turbolent) [inline runtime/interpreter/value_publickey.go:118]: The custom stringer is actually not needed, the default string formatting function produces the same output now

**Review by dsainati1 — APPROVED** [review]

File: runtime/stdlib/test_emulatorbackend.go [hunk 5]
```diff
@@ -322,14 +320,10 @@ func newTestAccountValue(
 	// Create address value
 	address := interpreter.NewAddressValue(nil, account.Address)
 
-	standardLibraryHandler := blockchain.StandardLibraryHandler()
```
Comment (SupunS) [inline runtime/stdlib/test_emulatorbackend.go:325]: @SupunS Note to myself to remove this function from the `Blockchain`

**Review by SupunS — APPROVED** [review]

File: runtime/stdlib/test_emulatorbackend.go (hunk 5 above)
Comment (SupunS) [reply runtime/stdlib/test_emulatorbackend.go:325]: https://github.com/onflow/cadence/issues/2898

**Pull request merged**